### DISCUSSION TITLE
ARC-0040: Accessing block metadata inside programs

### DISCUSSION THREAD

**Discussion by evan-schott** [ProvableHQ/ARCs#69] [ARCs]
---

arc: 0040
title: Accessing block metadata inside programs
authors: @evan-schott @d0cd
discussion: ARC-0040: Accessing block metadata inside programs
topic: Protocol
status: Draft
created: 2024-04-02

---

## Specification
Add support for the following opcodes in Aleo: `block.round`, `cumulative.weight`, `cumulative.proof_target`, `coinbase.target`, `proof.target`, `coinbase.target`, `block.timestamp`, `coinbase.timestamp`, `block.hash`, and `network.id`. The goal is for all [block metadata](https://github.com/AleoHQ/snarkVM/blob/925b3caed5e0b38f0bd7dfd6a1487a1812d301ae/ledger/block/src/header/metadata/mod.rs#L28-L51) to be accessible inside the finalize scope of deployed programs. Example: `block.timestamp 10u32` will return the timestamp of block `10`. 
## References
[Here](https://github.com/AleoHQ/snarkVM/pull/2419) is an example implementation of `network.id`.

**Comment by Delaw258** [comment]
nothing here

**Comment by zklim** [comment]
Yes definitely, using timestamp is more accurate than block heights to do time base calculations imo. Not sure about use-case of others yet.

**Reply by PNWBNW** [reply]
This would be awesome, along with having a bot that pays attention to UNIX, NIST, UNSO, ensuring minimal tolerance.
Creating dynamic block timestamps rather than just static.

**Comment by lxtrvn** [comment]
Yes this should already have been prt of the sytem.
